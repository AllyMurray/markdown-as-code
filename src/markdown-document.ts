import { heading } from './elements/heading.js';
import { DocumentSection, tryFindSection } from './sections/section.js';
import { tableOfContentsSection } from './sections/table-of-contents.js';

export interface MarkdownDocumentOptions {
  title: string;

  description?: string;

  content?: Array<DocumentSection>;

  /**
   * Include a table of contents in the output
   *
   * @default true
   */
  tableOfContents?: boolean;
}

export interface AddSection {
  section: DocumentSection;
  sortOrder?: number;
}

export class MarkdownDocument {
  private _tableOfContents: boolean;
  private _sortedSections: Array<DocumentSection> = [];

  private _title: string;

  constructor(private options: MarkdownDocumentOptions) {
    this._tableOfContents = options.tableOfContents ?? true;
    this._title = options.title;
    options.content?.forEach((section) => this.addSection(section));
    return this;
  }

  private get sections(): ReadonlyArray<DocumentSection> {
    return this._sortedSections;
  }

  public get title() {
    return this._title;
  }

  public addSection(section: DocumentSection) {
    this._sortedSections.push(section);
    return this;
  }

  public tryFindSection<S extends DocumentSection>(path: Array<string>): S {
    return tryFindSection(this.sections, path) as S;
  }

  private synthesizeSections(sections: ReadonlyArray<DocumentSection>) {
    const lines: Array<string> = [];
    for (const section of sections) {
      lines.push(section.synthesize());

      if (section.subSections.length > 0) {
        lines.push(...this.synthesizeSections(section.subSections));
      }
    }
    return lines;
  }

  public toString() {
    let tableOfContents;
    if (this._tableOfContents) {
      tableOfContents = tableOfContentsSection({
        sections: this.sections,
      }).synthesize();
    }

    const lines = [
      heading(1, this.options.title),
      this.options.description,
      tableOfContents,
      '',
    ];

    lines.push(...this.synthesizeSections(this.sections));

    return lines.filter(Boolean).join('\n\n');
  }
}

export const markdownDocument = (options: MarkdownDocumentOptions) =>
  new MarkdownDocument(options);
