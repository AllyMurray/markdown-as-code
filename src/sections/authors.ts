import { type ListOptions, ListSection } from './list-section.js';
import { link } from '../syntax/link.js';
import { mentionPerson } from '../syntax/mention.js';

interface Author {
  githubUsername: string;
}

export interface AuthorsOptions extends Partial<ListOptions> {}

export class Authors extends ListSection<Author> {
  constructor(options?: AuthorsOptions) {
    super({
      title: options?.title ?? 'Authors',
      type: 'Unordered',
      ...options,
    });
  }

  protected itemMapper(author: Author): string {
    return link(
      mentionPerson(author.githubUsername),
      new URL(author.githubUsername, 'https://www.github.com').toString()
    );
  }
}

export function authorsSection(options?: AuthorsOptions) {
  return new Authors(options);
}
