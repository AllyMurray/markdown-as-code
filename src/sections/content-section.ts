import { DocumentSection, DocumentSectionOptions } from './section.js';
import { heading } from '../syntax/heading.js';

interface ContentSectionOptions extends DocumentSectionOptions {
  content?: string;
}

export class ContentSection extends DocumentSection {
  private content: string;

  constructor(options: ContentSectionOptions) {
    super(options);
    this.content = options.content ?? '';
  }

  public appendContent(content: string) {
    this.content = this.content
      ? [this.content, '', content].join('\n')
      : content;
    return this;
  }

  protected synthesizeContent(): Array<string> {
    return [heading(this.headingLevel, this.title), '', this.content];
  }
}

export function contentSection(options: ContentSectionOptions) {
  return new ContentSection(options);
}
