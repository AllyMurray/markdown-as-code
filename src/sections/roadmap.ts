import { ListSection } from './list-section.js';

interface RoadmapItem {
  text: string;
}

export class Roadmap extends ListSection<RoadmapItem> {
  constructor(title?: string) {
    super(title ?? 'Roadmap', 'Unordered');
  }

  protected itemMapper(item: RoadmapItem): string {
    return item.text;
  }
}

export function roadmapSection(title?: string) {
  return new Roadmap(title);
}
