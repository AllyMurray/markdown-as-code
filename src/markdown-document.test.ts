import { markdownDocument } from './markdown-document.js';
import { Authors, authorsSection } from './sections/authors.js';
import { contentSection } from './sections/content-section.js';
import { roadmapSection } from './sections/roadmap.js';

describe('Markdown Document', () => {
  it('should synthesize a basic document', () => {
    const markdownDoc = markdownDocument({
      title: 'Test',
    });

    expect(markdownDoc.toString()).toMatchSnapshot();
  });

  it('should synthesize a document after adding a section', () => {
    const markdownDoc = markdownDocument({
      title: 'Test',
    }).addSection(roadmapSection({ items: [{ text: 'Item 1' }] }));

    expect(markdownDoc.toString()).toMatchSnapshot();
  });

  it('should synthesize a document after adding a custom section with a sub-section by setting rootSection', () => {
    const rootSection = contentSection({
      title: 'Root Section',
      content: 'Some markdown content',
    });

    contentSection({
      title: 'Sub Section',
      content: 'Some markdown sub content',
      parent: rootSection,
    });

    const markdownDoc = markdownDocument({
      title: 'Test',
      content: [rootSection],
    });

    expect(markdownDoc.toString()).toMatchSnapshot();
  });

  it('should synthesize a document after adding a custom section with a sub-section by using addSubSection', () => {
    const markdownDoc = markdownDocument({
      title: 'Test',
    }).addSection(
      contentSection({
        title: 'Root Section',
        content: 'Some markdown content',
      }).addSubSection(authorsSection().add({ githubUsername: 'jane-doe' })),
    );

    expect(markdownDoc.toString()).toMatchSnapshot();
  });

  it('should find a section by path', () => {
    const markdownDoc = markdownDocument({
      title: 'Test',
    });

    markdownDoc.addSection(
      authorsSection({ items: [{ githubUsername: 'john-smith' }] }),
    );

    const authorsSectionFromPath = markdownDoc.tryFindSection(['Authors']);

    expect(authorsSectionFromPath).toBeInstanceOf(Authors);
  });

  it('should return undefined when section is not found by path', () => {
    const markdownDoc = markdownDocument({
      title: 'Test',
    });

    markdownDoc.addSection(
      authorsSection({ items: [{ githubUsername: 'john-smith' }] }),
    );

    const authorsSectionFromPath = markdownDoc.tryFindSection(['Unknown']);

    expect(authorsSectionFromPath).toBeUndefined();
  });
});
