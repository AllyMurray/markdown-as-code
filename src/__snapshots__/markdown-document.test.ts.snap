// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown Document > should synthesize a basic document 1`] = `
"# Test

## Table of Contents
"
`;

exports[`Markdown Document > should synthesize a document after adding a custom section with a sub-section by setting rootSection 1`] = `
"# Test

## Table of Contents

- [Root Section](#root-section)
  - [Sub Section](#sub-section)

## Root Section

Some markdown content

### Sub Section

Some markdown sub content"
`;

exports[`Markdown Document > should synthesize a document after adding a custom section with a sub-section by using addSubSection 1`] = `
"# Test

## Table of Contents

- [Root Section](#root-section)
  - [Authors](#authors)

## Root Section

Some markdown content

### Authors

- [@jane-doe](https://www.github.com/jane-doe)"
`;

exports[`Markdown Document > should synthesize a document after adding a section 1`] = `
"# Test

## Table of Contents

- [Roadmap](#roadmap)

## Roadmap

- Item 1"
`;
